### DISCUSSION TITLE
Maps with all types as keys

### DISCUSSION THREAD

**Discussion by Wertzui123** [vlang/v#11927] [Ideas]
V currently only supports a few primitive types as keys in maps.
It would be useful if structs, arrays, ... could also be keys.

**Comment by JalonSolov** [comment]
How would this work?  What would be the use-case to have a single key be an array of ints, for example?

**Comment by Wertzui123** [comment]
I can't think of any use-cases for arrays, but structs are useful sometimes.

What do you mean by "how would this work"? It would work exactly the same as with other types.
I don't know how `map` is implemented in V, but if it's a hashmap it should be pretty straightforward.

**Comment by JalonSolov** [comment]
I'm just having a hard time wrapping my head around why anyone would need to have a struct as a key in a map.  There are so many things that could make it work incorrectly, depending on instances, pointers, etc.

But as you say, it does use hashes, so it would probably work at least some of the time.  :-\

**Comment by Wertzui123** [comment]
For example, when writing plugins for some kind of service, maps with struct keys can be a replacement for fields that you can't declare in the plugin, because the struct is declared inside the main service.

**Comment by JalonSolov** [comment]
I would never do plugins that way... I would only want to load a plugin once.  The `struct` could be passed to the plugin's `init` routine to set it up, but it's just wasteful to use the entire struct as the key in a map.  This would also allow calling the same plugin multiple times with different `structs`, instead of having to load multiple instances of the same plugin just because the struct changed.

At any rate... to each his own, I guess.  I'll stop worrying about it now.  It's up to someone else whether or not to implement this.

**Comment by Wertzui123** [comment]
I meant that e.g. your service passes a user object to the plugin. The plugin then wants to do something if the user earlier turned on a feature of the plugin in their personal settings.
So, it checks:
```v
if user_feature_map[user] == true{
```
(This design is of course only practical if one user only has one object assigned ever, so basically, the service has to store everything in the RAM all the time...)

Anyway, I'm sure there are some scenarios where this is useful, so I agree with you to just leave this open so someone can someday implement this.

**Comment by mcastorina** [comment]
For reference, Rust's HashMaps allow any type to be a key as long as it satisfies two constrainst:

1. The type can be compared for equality
2. The type can be hashed

In Rust these are the `Eq` and `Hash` traits, which can be manually implemented for custom types. This means you can make `Eq` and `Hash` operate on specific struct attributes, rather than everything, which could be useful.

The equivalent in V would be a `Hashable` interface.

**Comment by Wertzui123** [comment]
By the way, I have written a hashmap implementation that works like Rust's HashMaps, as described above: https://github.com/Wertzui123/HashMap

**Comment by MiroloTech** [comment]
Here is a usecase, which Coooouuuld be implemented as two maps, but this just allows for way cleaner and prettier code:
![image](https://github.com/user-attachments/assets/176cb08e-e2ff-403e-b4e3-76a09ed87630)
I don't want to turn everything into a hashmap, and just use what I have available.

**Comment by Wajinn** [comment]
This is also related to the feature request of "[Array as a map key](https://github.com/vlang/v/issues/22992)".
